pkg/gardenlet/controller/shoot/shoot/reconciler.go

```diff
@@ -804,7 +804,11 @@ func (r *Reconciler) patchShootStatusOperationError(
 }
 
 func (r *Reconciler) shootHasBastions(ctx context.Context, shoot *gardencorev1beta1.Shoot) (bool, error) {
-	return kubernetesutils.ResourcesExist(ctx, r.GardenClient, operationsv1alpha1.SchemeGroupVersion.WithKind("BastionList"), client.MatchingFields{operations.BastionShootName: shoot.Name})
```

Comment:
> However, using `PartialObjectMetadata` in combination with a client cache and field selectors doesn't work well

Isn't this a problem? Don't we use this combination at many other places? TBH, I don't find it very obvious and probably forget about it again after the weekend. What can we do to support this?

Reply:
We usually use `kubernetesutils.ResourcesExist` with selectors which target information in `Metadata`, e.g. `namespace` or `labels`.

> What can we do to support this?

There is nothing we can do to support the current implementation of `kubernetesutils.ResourcesExist`. With `PartialObjectMetadataList` you can't instruct an indexer to [index](https://github.com/gardener/gardener/blob/80feac852da0fcdbee16167cde509f067673e766/pkg/api/indexer/gardener_operations.go#L29-L33) a field from any other part than `Metadata`.

> TBH, I don't find it very obvious and probably forget about it again after the weekend.

I see some options:
1. Take the current changes of this PR -> Proper usage is documented in `kubernetesutils.ResourcesExist` and if misused, the client will directly return an error (caught by tests).
2. Use the `APIReader` for `kubernetesutils.ResourcesExist` -> also not very obvious, no protection from misusage.
3. `kubernetesutils.ResourcesExist` used the object list directly instead of `PartialObjectMetadataList`.

I prefer option 3. WDYT?

Reply:
> There is nothing we can do to support the current implementation of `kubernetesutils.ResourcesExist`. With `PartialObjectMetadataList` you can't instruct an indexer to [index](https://github.com/gardener/gardener/blob/80feac852da0fcdbee16167cde509f067673e766/pkg/api/indexer/gardener_operations.go#L29-L33) a field from any other part than `Metadata`.

LOL :) Of course, sorry. Makes perfect sense 😄 

What about keeping the doc, and additionally check in `kubernetesutils.ResourcesExist` whether the passed `listOpts` contain field selectors? In this case, we could at least return an error?

Reply:
After looking at [3a795ac](https://github.com/gardener/gardener/pull/8561/commits/3a795ac97b0f0c24a604a811b326fd6ff3b9bf6d), another idea came to my mind: Instead of returning an error when somebody passes a field selector to `kubernetesutils.ResourcesExist`, wouldn't it be nice if the function then simply not uses `PartialObjectMetadataList` but the real object list (via `unstructured.Unstructured`)? Then the caller does not need to take care about these implementation details and the function optimizes internally based on the passed list options. WDYT? Sorry for the back-and-forth, this only came to my mind now after seeing that you had to adapt a few other places were field selectors were used as well.

Reply:
It's better to avoid working with `unstructured.Unstructured` since the client creates and maintains another informer instance which can become expensive for some types in bigger environments (e.g. `shoots`).

I found another approach to achieve what you suggested, though - see [807b5f6](https://github.com/gardener/gardener/pull/8561/commits/807b5f64a40989393e126652d685d9be6ee692cf).